### PR TITLE
daemon: add capability to wait for endpoint to be in "ready" state if specified via EndpointChangeRequest

### PR DESCRIPTION
**Pull request by ianvernon**
This allows for different behavior depending upon the orchestration system being used in tandem with Cilium for calls to /PUT for an endpoint. Different orchestration systems handle bootstrapping / initialization of container / endpoints differently, and as such whether API calls should synchronously wait for endpoint regeneration to occur must be configurable.

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #4409

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4413)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
@tgraf CI passed doing this by default for the CNI plugin.

**Comment by tgraf** [comment]
Awesome. Let's ship it and backport it as well.

File: pkg/endpoint/endpoint_test.go [hunk 1]
```diff
@@ -221,19 +221,7 @@ func (s *EndpointSuite) TestEndpointState(c *C) {
 	e.SetDefaultOpts(nil)
 	defer e.Mutex.Unlock()
 
-	e.state = StateCreating
```
Comment (manalibhutiyani) [inline pkg/endpoint/endpoint_test.go]: Ditto.

File: daemon/endpoint.go [hunk 2]
```diff
@@ -211,6 +211,44 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
 	if err != nil {
 		apierror.Error(code, err)
 	}
+
+	// Wait for endpoint to be ready if specified in API call.
+	if params.Endpoint.SyncBuildEndpoint {
```
Comment (manalibhutiyani) [inline daemon/endpoint.go]: Quick question. Please correct me if I am wrong. Will this be a problem if we try to create a docker container without labels. Since in that case, it will be EndpointStateCreating /StateWaitingForIdentity(Not sure how this behavior has changed with the new Init identity)? I believe, in docker container creation, there is a transition from EndpointStateCreating to other states which is spilt between in PUT (creation) and PATCH. Will that be a problem, now that we are synchronously waiting for the state to be `ready` in PUT itself?


This is fine case of K8s which does not have the split between PUT/PATCH but here, will we be waiting synchronously for the state change ?

File: pkg/endpoint/endpoint.go [hunk 3]
```diff
@@ -1873,11 +1863,6 @@ func (e *Endpoint) SetStateLocked(toState, reason string) bool {
 	// Validate the state transition.
 	fromState := e.state
 	switch fromState { // From state
-	case StateCreating:
```
Comment (manalibhutiyani) [inline pkg/endpoint/endpoint.go]: We should not be removing `StateCreating` for backward compatibility. The earlier api still allows `StateCreating` 
CC @rlenglet I asked Romain the same question in the other PR #3863 I was fixing, and he asked me to not remove this state to maintain back ward api compatibility.

File: daemon/endpoint.go (hunk 2 above)
Comment (tgraf) [reply daemon/endpoint.go]: The flag is only enabled via CNI so it will not affect libnetwork setups

File: pkg/endpoint/endpoint.go (hunk 3 above)
Comment (tgraf) [reply pkg/endpoint/endpoint.go]: Let's isolate the change of removing the Creating state and dedicate a separate PR for that. I don't understand the implications yet.

File: daemon/endpoint.go [hunk 4]
```diff
@@ -211,6 +211,44 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
...
+		log.Debug("synchronously waiting for endpoint to be in %s state", endpoint.StateReady)
```
Comment (tgraf) [inline daemon/endpoint.go]: We should include at least an endpoint ID in the log message here

File: daemon/endpoint.go [hunk 5]
```diff
@@ -211,6 +211,44 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
...
+
+		e, err := endpointmanager.Lookup(params.ID)
+		if err != nil {
+			return apierror.Error(PutEndpointIDFailedCode, err)
```
Comment (tgraf) [inline daemon/endpoint.go]: We have to remove the endpoint again this case. The API must be consistent and either return success with the endpoint still inserted or an error without leaving behind an endpoint.

File: daemon/endpoint.go [hunk 6]
```diff
@@ -211,6 +211,44 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
...
+			select {
+			case <-ticker.C:
+				e.Mutex.Lock()
+				epState := e.GetStateLocked()
+				e.Mutex.Unlock()
+				if epState == endpoint.StateReady {
+					return nil
+				}
+			case <-timeout:
+				return apierror.Error(PutEndpointIDFailedCode, fmt.Errorf("endpoint did not synchronously regenerate after timeout"))
```
Comment (tgraf) [inline daemon/endpoint.go]: Same here, endpoint must be removed again

**Review by tgraf — DISMISSED** [review]

File: pkg/endpoint/endpoint.go (hunk 3 above)
Comment (ianvernon) [reply pkg/endpoint/endpoint.go]: Sure, will do.

File: pkg/endpoint/endpoint_test.go (hunk 1 above)
Comment (ianvernon) [reply pkg/endpoint/endpoint_test.go]: Removed from this upload.

File: daemon/endpoint.go (hunk 4 above)
Comment (ianvernon) [reply daemon/endpoint.go]: Done.

File: daemon/endpoint.go (hunk 5 above)
Comment (ianvernon) [reply daemon/endpoint.go]: Done.

File: daemon/endpoint.go (hunk 6 above)
Comment (ianvernon) [reply daemon/endpoint.go]: Done.

**Comment by ianvernon** [comment]
test-me-please

**Review by manalibhutiyani — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Hit a “go vet” issue - mistakenly duplicated the check for disconnecting / disconnected state. Will fix and retest.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by raybejjani** [comment]
I realise this ship has sailed. It would have been better to place this logic in the client library we provide.
We now have two different behaviours in the API. I'm surprised that we need to delete the endpoint on fail only in the sync case, actually, since this suggests that we left deletions up to the client. In any case, this toggle means we now have two response mechanisms a client needs to support and understand, one that is synchronous *and* transactional (it seems, from the delete) and another that is loosely async (but is really just underspecified). I had intended for revision numbers to go into Endpoint, and that all clients would async poll as needed (or, more generically, have a blocking wait for revision X for Endpoint Y API they can use).
More generally, and it should never happen in this case, implementing the block serverside means the server shoulders the load of many concurrently blocking requests. In this case, these calls come from the same machine, so it's fine.

**Comment by tgraf** [comment]
The deletion is a must, we should not leave any state behind if the API to create an object fails.

I'm fine to move this to the client side. This is unreleased and we can still change it. What benefits do you see in doing this on the client side? It means that we have to handle deletion from the client as well and we will rely on clients to not leave behind state on failure.

**Comment by raybejjani** [comment]
Well, you said that deletion is required in all cases? or did I understand the implication? (If I haven't do we need to move it out of the if in this PR?)

To me, the root issue is whether the API is consistent in it's basic behaviour. Will we end up with synchronous calls for all API calls? An example would be you change an endpoint config and want to wait until it is in the datapath (or until it's "complete", whatever that might mean). We can say, no, this is not supported, but then why is this other case supported? (obviously because it's more important to wait but I'm the caller of the API, I don't want to know about implementation details). Similarly, when a client adds a policy, it must then poll to get the endpoint statuses. Why doesn't that have a "wait for all affected endpoints to implement this policy"? Somewhat relatedly, a strictly async API reflects the normal way events propagate in the agent.

I don't think the scaling argument matters that much here, but you can imagine batching a wait call for multiple endpoints from the client-side. Honestly, though, that isn't going to be so different from the client just making a bunch of blocking requests.

The error return path here is also a little overloaded, since we return an error on the endpoint create's timeouts and failures as well. These are probably 5xxs so it's consistent, but it does mean that in some modes the success of the endpoint create is part of the response information, instead of this only being the request's success/fail. This is quite intuitive, I think, but is there an equivalent way to do this in the fire-and-forget case? I don't think so, since if the endpoint doesn't create it would be deleted and there is no record of it :/

The one other value in adding this to the client library is that it becomes an example on how to poll our API to get information. If someone wants to use the library, perfect. If they need a more complex setup they have a ready made starting snippet. In this case, only someone adding a new orchestration system would be interested anyway. I guess I also see some value in forcing ourselves to support the async polling case, since it forces us to have correct bookkeeping inside the agent, but that isn't terribly important either.

An argument, among many, for this method is that it simplifies clients enough that a trivial curl can be used. This is handy since scripts etc. are probably annoying to write with correct retries.

Anyway, I think this is ultimately a stylistic thing. I prefer a simple API and more complex clients, but this case especially is hard because the create seems to integral to this request.

**Comment by tgraf** [comment]
Good argumentation. I think we have two major arguments here:

* API consistency to not expect different logic for difference API calls that involve endpoint state changes. This would clearly speak to move this logic into the client. The counter is code duplication which can be resolved by having this logic available in `pkg/client`
* API usage simplicity:  The extensive logic embedded into `kubectl` vs what is in the apiserver has made me a bit cautious as the overall learning after the years was that client side logic was not beneficial in the end. There is one key question that matters here: Do we need to know whether the client will consider the request to have been successful or not? In particular if we want to hook additional logic after the initial creation of the endpoint. I think this question is very hard to answer right now.

In conclusion: I understand and agree with the consistency argument. While we can still do the server side logic if we chose the client side logic, we can't undo the server side logic after we have put it in. Therefore, let's move it to the client and see how that goes and if needed, introduce the server side logic to gain better control later on.

**Comment by raybejjani** [comment]
> The extensive logic embedded into kubectl vs what is in the apiserver has made me a bit cautious as the overall learning after the years was that client side logic was not beneficial in the end. 

I don't know enough about kubectl to comment but, to be clear, I'm arguing over how to do sync/async vs anything more complex (e.g. the create should all happen server-side).

> Do we need to know whether the client will consider the request to have been successful or not?

In this case, you mean we would then do something different for that client? (e.g. the CNI plugin would get an error if it reuses some unique ID for a new pod/container because we already saw it.. or something). If so, yeah, I don't have a solution. All schemes I can think of are a bit too complex to justify themselves and it would be better to do the serverside block.
